edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.AuthAgent/AuthAgentListener.cs

```diff
@@ -0,0 +1,532 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Hub.AuthAgent
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Net;
+    using System.Security.Cryptography.X509Certificates;
+    using System.Text;
+    using System.Threading;
+    using System.Threading.Tasks;
+    
+    using Microsoft.Azure.Devices.Edge.Hub.Core;
+    using Microsoft.Azure.Devices.Edge.Hub.Core.Identity;
+    using Microsoft.Azure.Devices.Edge.Util;
+
+    using Microsoft.Extensions.Logging;
+    using Newtonsoft.Json;
+
+    public class AuthAgentListener : IProtocolHead
+    {
+        private readonly IAuthenticator authenticator;
+        private readonly IUsernameParser usernameParser;
+        private readonly IClientCredentialsFactory clientCredentialsFactory;
+        private readonly string listeningAddress;
+
+        private HttpListener listener;
+
+        private Task listenerTask;
+        private bool exitListenerLoop;
+
+        private const int maxInBufferSize = 32 * 1024;
+        private static readonly TimeSpan requestTimeout = TimeSpan.FromSeconds(5);
+
+        public AuthAgentListener(IAuthenticator authenticator, IUsernameParser usernameParser, IClientCredentialsFactory clientCredentialsFactory, string listeningAddress)
+        {
+            this.authenticator = Preconditions.CheckNotNull(authenticator, nameof(authenticator));
+            this.usernameParser = Preconditions.CheckNotNull(usernameParser, nameof(usernameParser));
+            this.clientCredentialsFactory = Preconditions.CheckNotNull(clientCredentialsFactory, nameof(clientCredentialsFactory));
+            this.listeningAddress = Preconditions.CheckNotNull(listeningAddress, nameof(listeningAddress));
+        }
+
+        public string Name => "AUTH";
+
+        public Task StartAsync()
+        {
+            Events.Starting();
+            this.Start();
+            Events.Started();
+
+            return Task.CompletedTask;
+        }
+
+        public async Task CloseAsync(CancellationToken token)
+        {
+            Events.Closing();
+
+            var currentListener = Interlocked.Exchange(ref this.listener, null);
+            if (currentListener == null)
+            {
+                Events.ClosedWhenNotRunning();
+                return;
+            }
+
+            // this disposes the instance making the listener loop exit
+            Volatile.Write(ref exitListenerLoop, true);
+            currentListener.Close();
+
+            await this.listenerTask;
+            this.listenerTask = null;
+
+            Events.Closed();
+        }
+
+        public void Dispose() => this.CloseAsync(CancellationToken.None).Wait();
+
+        private void Start()
+        {
+            var newListener = new HttpListener();
```

Comment:
HttpListener needs Admin on Windows. Better approach would be to use the existing HttpProtocol head. We can open up port 80 (or any other port we want), and limit traffic on that port to the Authentication controller. Of course, that port would not be exposed out.

Reply:
I had to make a very similar change for the metrics endpoint: https://github.com/Azure/iotedge/compare/master...lfitchett:asp.net_metrics

Reply:
Changed using aspnet. Created a separate host, so other settings (authentications, maybe middlewares) will not interfere. Also the auth endpoint is extremely simple and now it can live its own life.